### DISCUSSION TITLE
关于OpenSSL的req子命令的疑惑 | Questions About the OpenSSL req Subcommand

### DISCUSSION THREAD

**Discussion by Meowilla** [openssl/openssl#25335] [Q&A] [answered]
关于OpenSSL的openssl req命令，我查看了它的帮助文档，对于里面的内容有几处问题，以下是我所记录的相关问题

1. openssl req的-key选项，帮助文档内描述`对于证书签名，此选项被-CA选项所取代 `，被-CA选项所取代？我测试并没有这种情况
2. openssl req的-key选项，帮助文档内描述 `除非给出了-in，否则将在新证书或证书请求中放置相应的公钥，从而产生自签名`，这应在指定了-x509选项的情况中才会发生自签名，或者是我的理解问题
3. openssl req的-x509选项，帮助文档内描述 `它隐含了-CA选项 `，在-CA的帮助内也是 `它隐含了-x509”选项 `，-CA选项应该需要一个参数，但我在使用-x509时并不需要指定参数，我也并不理解这是什么意思，-CA和-x509应该是一个用来签署证书，一个用来生成自签证书

这些问题是文档的编写错误吗？还是我理解错了？

Here is the English translation:

---

**Questions About the OpenSSL `openssl req` Command**

I have reviewed the help documentation for the `openssl req` command and have encountered a few issues that I would like to clarify. Below are the specific concerns I have noted:

1. For the `-key` option in `openssl req`, the documentation states that "for certificate signing, this option is replaced by the `-CA` option." Replaced by the `-CA` option? I did not observe this behavior in my testing.

2. For the -key option in openssl req, the documentation states that "unless the -in option is provided, the corresponding public key will be placed in the new certificate or certificate request, resulting in a self-signed certificate." This self-signing should only happen when the -x509 option is specified, or is it a misunderstanding on my part?

3. For the `-x509` option in `openssl req`, the documentation states that "it implies the `-CA` option." Similarly, in the help for the `-CA` option, it states that "it implies the `-x509` option." The `-CA` option should require a parameter, but I don’t need to specify one when using `-x509`, and I don’t understand what this means. The `-CA` option is supposed to be used for signing certificates, while the `-x509` option is used to generate a self-signed certificate.

Are these documentation errors, or am I misunderstanding something?

**Comment by t8m** [comment, marked as answer]
> 1. For the `-key` option in `openssl req`, the documentation states that "for certificate signing, this option is replaced by the `-CA` option." Replaced by the `-CA` option? I did not observe this behavior in my testing.

Not replaced but overridden. If you use `-CA`, the `-key` option will be ignored.

> 2. For the -key option in openssl req, the documentation states that "unless the -in option is provided, the corresponding public key will be placed in the new certificate or certificate request, resulting in a self-signed certificate." This self-signing should only happen when the -x509 option is specified, or is it a misunderstanding on my part?

Well, certificate requests are always self-signed. The sentence in the documentation could be probably made more clear.

> 3. For the `-x509` option in `openssl req`, the documentation states that "it implies the `-CA` option." Similarly, in the help for the `-CA` option, it states that "it implies the `-x509` option." The `-CA` option should require a parameter, but I don’t need to specify one when using `-x509`, and I don’t understand what this means. The `-CA` option is supposed to be used for signing certificates, while the `-x509` option is used to generate a self-signed certificate.

No, the docs say that `-x509` is implied by `-CA`. That means, if you specify `-CA /some/path/to/CA/certificate`, you do not need to specify `-x509` in addition to that to produce a signed certificate. I do not see any issue here.

**Reply by Meowilla** [reply]
Thankyou very much!
